File: problems/1209.0_Remove_All_Adjacent_Duplicates_in_String_II.py
'''
stack

Runtime: 212 ms, faster than 26.34% of Python3 online submissions for Remove All Adjacent Duplicates in String II.
Memory Usage: 18.6 MB, less than 75.09% of Python3 online submissions for Remove All Adjacent Duplicates in String II.
'''
class Solution:
    def removeDuplicates(self, s: str, k: int) -> str:
        # [char, count]
        stack = []
        for i, ch in enumerate(s):
            if not stack or stack[-1][0] != ch:
                stack.append([ch, 1])
            else:
                stack[-1][1] += 1
                if stack[-1][1] == k:
                    stack.pop()
        
        return ''.join(ch * cnt for ch, cnt in stack)
            

File: cpp/1209.0_Remove_All_Adjacent_Duplicates_in_String_II.cpp
/*
Runtime: 15 ms, faster than 93.44% of C++ online submissions for Remove All Adjacent Duplicates in String II.
Memory Usage: 11 MB, less than 30.37% of C++ online submissions for Remove All Adjacent Duplicates in String II.
*/
class Solution {
public:
    string removeDuplicates(string s, int k) {
        int i = 0, n = s.length();
        vector<int> count(n);
        for (int j = 0; j < n; ++j, ++i) {
            s[i] = s[j];
            count[i] = i > 0 && s[i - 1] == s[j] ? count[i - 1] + 1 : 1;
            if (count[i] == k) i -= k;
        }
        return s.substr(0, i);
    }
};
